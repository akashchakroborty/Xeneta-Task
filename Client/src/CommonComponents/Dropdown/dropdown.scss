@import '../../App/app.variables.scss';

.dropdown {
  font-size: $primary-font-size;
  font-weight: bold;
  color: $primary-color;
  text-decoration: underline;
  text-decoration-color: $primary-color;
  text-underline-offset: 0.5em;
  .ant-select-arrow {
    color: $primary-color;
  }
}
