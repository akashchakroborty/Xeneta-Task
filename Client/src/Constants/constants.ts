export const SELECT_ORIGIN = 'Select Origin';
export const SELECT_DESTINATION = 'Select Destination';
export const EMPTY_STRING = '';
export const DEFAULT_PORTS = [];
export const DEFAULT_RATES = [];
export const DEFAULT_LOADING = false;
