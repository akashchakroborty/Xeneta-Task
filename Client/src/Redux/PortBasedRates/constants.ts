export const SET_PORTS_SUCCESS = 'SET_PORTS_SUCCESS';
export const SET_PORTS_FAILURE = 'SET_PORTS_FAILURE';
export const GET_PORTS_LOADING = 'GET_PORTS_LOADING';
export const UPDATE_ORIGIN = 'UPDATE_ORIGIN';
export const UPDATE_DESTINATION = 'UPDATE_DESTINATION';
