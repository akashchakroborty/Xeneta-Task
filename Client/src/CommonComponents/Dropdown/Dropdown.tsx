import React, { useEffect } from 'react';
import { Select } from 'antd';
import { Ports } from '../../Redux/Types/portBasedRatesTypes';
import { SELECT_DESTINATION, SELECT_ORIGIN } from '../../Constants/constants';

import './dropdown.scss';

export interface DropdownProps {
  getPorts: () => void;
  updateOrigin: (origin: string) => void;
  updateDestination: (destination: string) => void;
  ports: Ports[];
  defaultOption: string;
  isPortsLoading: boolean;
}

const { Option } = Select;

const DropdownComponent: React.FC<DropdownProps> = ({
  getPorts,
  updateOrigin,
  updateDestination,
  ports,
  defaultOption,
  isPortsLoading,
}) => {
  const handlePortSelect = (value: string) => {
    if (defaultOption === SELECT_ORIGIN) {
      updateOrigin(value);
    }
    if (defaultOption === SELECT_DESTINATION) {
      updateDestination(value);
    }
  };

  useEffect(() => {
    if (ports.length === 0) {
      getPorts();
    }
  }, []);

  return (
    <div>
      <Select
        defaultValue={defaultOption}
        bordered={false}
        className="dropdown"
        onChange={handlePortSelect}
        loading={isPortsLoading}
        showSearch
      >
        {ports.map((port: Ports) => (
          <Option key={port.code} value={port.code}>
            {port.name} ({port.code})
          </Option>
        ))}
      </Select>
    </div>
  );
};

export default DropdownComponent;
