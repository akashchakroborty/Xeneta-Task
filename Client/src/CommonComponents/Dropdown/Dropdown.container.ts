import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { getPorts } from '../../Redux/PortBasedRates/actionCreators';
import { selectPorts } from '../../Redux/PortBasedRates/selector';
import { AppState } from '../../Redux/store';
import { PortBasedRatesActionTypes } from '../../Redux/Types/portBasedRatesActionTypes';
import Dropdown from './Dropdown';

const mapStateToProps = (state: AppState) => ({
    ports: selectPorts(state),
  });

const mapDispatchToProps = (dispatch: ThunkDispatch<AppState, void, PortBasedRatesActionTypes>) => {
  return {
    getPorts: () => {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      dispatch(getPorts());
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Dropdown);
