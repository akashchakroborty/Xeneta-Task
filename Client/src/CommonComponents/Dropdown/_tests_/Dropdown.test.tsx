import React from 'react';
import { shallow } from 'enzyme';
import Dropdown, { DropdownProps } from '../Dropdown';
import { SELECT_DESTINATION, SELECT_ORIGIN } from '../../../Constants/constants';
import { MOCK_PORTS_DATA } from '../../../Constants/testData';

describe('should test the Dropdown Component', () => {
  const mockProps: DropdownProps = {
    getPorts: jest.fn(),
    updateOrigin: jest.fn(),
    updateDestination: jest.fn(),
    ports: MOCK_PORTS_DATA,
    defaultOption: SELECT_ORIGIN,
    isPortsLoading: false,
  };
  test('should render the Component', () => {
    const component = shallow(<Dropdown {...mockProps} />);
    expect(component).toMatchSnapshot();
  });
  test('should test handlePortSelect function', () => {
    const wrapper = shallow(<Dropdown {...mockProps} />);
    wrapper.find('.dropdown').simulate('change', 'mockValue1');
    expect(mockProps.updateOrigin).toHaveBeenCalledWith('mockValue1');
  });
  test('should test handlePortSelect function for Destination', () => {
    const newMockProps: DropdownProps = { ...mockProps, defaultOption: SELECT_DESTINATION };
    const wrapper = shallow(<Dropdown {...newMockProps} />);
    wrapper.find('.dropdown').simulate('change', 'mockValue1');
    expect(mockProps.updateDestination).toHaveBeenCalledWith('mockValue1');
  });
});
