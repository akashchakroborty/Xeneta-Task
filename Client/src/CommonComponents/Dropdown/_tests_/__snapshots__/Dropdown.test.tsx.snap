// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the Dropdown Component should render the Component 1`] = `
<div>
  <ForwardRef(InternalSelect)
    bordered={false}
    className="dropdown"
    defaultValue="Select Origin"
    loading={false}
    onChange={[Function]}
    showSearch={true}
  >
    <Option
      key="NOOSL"
      value="NOOSL"
    >
      Oslo
       
      (
      NOOSL
      )
    </Option>
    <Option
      key="CNSGH"
      value="CNSGH"
    >
      Shanghai
       
      (
      CNSGH
      )
    </Option>
    <Option
      key="CNSTG"
      value="CNSTG"
    >
      Shantou
       
      (
      CNSTG
      )
    </Option>
    <Option
      key="NLRTM"
      value="NLRTM"
    >
      Rotterdam
       
      (
      NLRTM
      )
    </Option>
    <Option
      key="DEHAM"
      value="DEHAM"
    >
      Hamburg
       
      (
      DEHAM
      )
    </Option>
    <Option
      key="GBFXT"
      value="GBFXT"
    >
      Felixstowe
       
      (
      GBFXT
      )
    </Option>
    <Option
      key="USNYC"
      value="USNYC"
    >
      New York
       
      (
      USNYC
      )
    </Option>
  </ForwardRef(InternalSelect)>
</div>
`;
